// @angular/compiler-cli/src/ngtsc/typecheck/diagnostics/src/diagnostic.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/typecheck/diagnostics/src/diagnostic", ["require", "exports", "tslib", "typescript"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.isTemplateDiagnostic = exports.makeTemplateDiagnostic = void 0;
    var tslib_1 = require("tslib");
    var ts = require("typescript");
    /**
     * Constructs a `ts.Diagnostic` for a given `ParseSourceSpan` within a template.
     */
    function makeTemplateDiagnostic(templateId, mapping, span, category, code, messageText, relatedMessages) {
        var e_1, _a, e_2, _b;
        if (mapping.type === 'direct') {
            var relatedInformation = undefined;
            if (relatedMessages !== undefined) {
                relatedInformation = [];
                try {
                    for (var relatedMessages_1 = tslib_1.__values(relatedMessages), relatedMessages_1_1 = relatedMessages_1.next(); !relatedMessages_1_1.done; relatedMessages_1_1 = relatedMessages_1.next()) {
                        var relatedMessage = relatedMessages_1_1.value;
                        relatedInformation.push({
                            category: ts.DiagnosticCategory.Message,
                            code: 0,
                            file: relatedMessage.sourceFile,
                            start: relatedMessage.start,
                            length: relatedMessage.end - relatedMessage.start,
                            messageText: relatedMessage.text,
                        });
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (relatedMessages_1_1 && !relatedMessages_1_1.done && (_a = relatedMessages_1.return)) _a.call(relatedMessages_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
            // For direct mappings, the error is shown inline as ngtsc was able to pinpoint a string
            // constant within the `@Component` decorator for the template. This allows us to map the error
            // directly into the bytes of the source file.
            return {
                source: 'ngtsc',
                code: code,
                category: category,
                messageText: messageText,
                file: mapping.node.getSourceFile(),
                componentFile: mapping.node.getSourceFile(),
                templateId: templateId,
                start: span.start.offset,
                length: span.end.offset - span.start.offset,
                relatedInformation: relatedInformation,
            };
        }
        else if (mapping.type === 'indirect' || mapping.type === 'external') {
            // For indirect mappings (template was declared inline, but ngtsc couldn't map it directly
            // to a string constant in the decorator), the component's file name is given with a suffix
            // indicating it's not the TS file being displayed, but a template.
            // For external temoplates, the HTML filename is used.
            var componentSf = mapping.componentClass.getSourceFile();
            var componentName = mapping.componentClass.name.text;
            // TODO(alxhub): remove cast when TS in g3 supports this narrowing.
            var fileName = mapping.type === 'indirect' ?
                componentSf.fileName + " (" + componentName + " template)" :
                mapping.templateUrl;
            // TODO(alxhub): investigate creating a fake `ts.SourceFile` here instead of invoking the TS
            // parser against the template (HTML is just really syntactically invalid TypeScript code ;).
            // Also investigate caching the file to avoid running the parser multiple times.
            var sf = ts.createSourceFile(fileName, mapping.template, ts.ScriptTarget.Latest, false, ts.ScriptKind.JSX);
            var relatedInformation = [];
            if (relatedMessages !== undefined) {
                try {
                    for (var relatedMessages_2 = tslib_1.__values(relatedMessages), relatedMessages_2_1 = relatedMessages_2.next(); !relatedMessages_2_1.done; relatedMessages_2_1 = relatedMessages_2.next()) {
                        var relatedMessage = relatedMessages_2_1.value;
                        relatedInformation.push({
                            category: ts.DiagnosticCategory.Message,
                            code: 0,
                            file: relatedMessage.sourceFile,
                            start: relatedMessage.start,
                            length: relatedMessage.end - relatedMessage.start,
                            messageText: relatedMessage.text,
                        });
                    }
                }
                catch (e_2_1) { e_2 = { error: e_2_1 }; }
                finally {
                    try {
                        if (relatedMessages_2_1 && !relatedMessages_2_1.done && (_b = relatedMessages_2.return)) _b.call(relatedMessages_2);
                    }
                    finally { if (e_2) throw e_2.error; }
                }
            }
            relatedInformation.push({
                category: ts.DiagnosticCategory.Message,
                code: 0,
                file: componentSf,
                // mapping.node represents either the 'template' or 'templateUrl' expression. getStart()
                // and getEnd() are used because they don't include surrounding whitespace.
                start: mapping.node.getStart(),
                length: mapping.node.getEnd() - mapping.node.getStart(),
                messageText: "Error occurs in the template of component " + componentName + ".",
            });
            return {
                source: 'ngtsc',
                category: category,
                code: code,
                messageText: messageText,
                file: sf,
                componentFile: componentSf,
                templateId: templateId,
                start: span.start.offset,
                length: span.end.offset - span.start.offset,
                // Show a secondary message indicating the component whose template contains the error.
                relatedInformation: relatedInformation,
            };
        }
        else {
            throw new Error("Unexpected source mapping type: " + mapping.type);
        }
    }
    exports.makeTemplateDiagnostic = makeTemplateDiagnostic;
    function isTemplateDiagnostic(diagnostic) {
        return diagnostic.hasOwnProperty('componentFile') &&
            ts.isSourceFile(diagnostic.componentFile);
    }
    exports.isTemplateDiagnostic = isTemplateDiagnostic;
});
